renderdoc/driver/d3d12/d3d12_shaderdebug.cpp

```diff
@@ -131,8 +131,645 @@ bool D3D12DebugAPIWrapper::CalculateSampleGather(
+                     strSwizzle + ";";
+    sampleProgram += "\n}\n";
+  }
+
+  // Create VS/PS to fetch the sample
+  ID3DBlob *vsBlob = NULL;
+  ID3DBlob *psBlob = NULL;
+  UINT flags = D3DCOMPILE_DEBUG | D3DCOMPILE_WARNINGS_ARE_ERRORS | D3DCOMPILE_DEBUG_NAME_FOR_SOURCE;
+  if(m_pDevice->GetShaderCache()->GetShaderBlob(vsProgram.c_str(), "main", flags, "vs_5_0",
+                                                &vsBlob) != "")
+  {
+    RDCERR("Failed to create shader to extract inputs");
+    return false;
+  }
+  if(m_pDevice->GetShaderCache()->GetShaderBlob(sampleProgram.c_str(), "main", flags, "ps_5_0",
+                                                &psBlob) != "")
+  {
+    RDCERR("Failed to create shader to extract inputs");
+    SAFE_RELEASE(vsBlob);
+    return false;
+  }
+
+  // Create a PSO with our VS/PS and all other state from the original event
+  D3D12RenderState &rs = m_pDevice->GetQueue()->GetCommandData()->m_RenderState;
+  D3D12RenderState prevState = rs;
+  WrappedID3D12PipelineState *origPSO =
+      m_pDevice->GetResourceManager()->GetCurrentAs<WrappedID3D12PipelineState>(rs.pipe);
+
+  D3D12_EXPANDED_PIPELINE_STATE_STREAM_DESC pipeDesc;
+  origPSO->Fill(pipeDesc);
+  pipeDesc.VS.BytecodeLength = vsBlob->GetBufferSize();
+  pipeDesc.VS.pShaderBytecode = vsBlob->GetBufferPointer();
+  pipeDesc.PS.BytecodeLength = psBlob->GetBufferSize();
+  pipeDesc.PS.pShaderBytecode = psBlob->GetBufferPointer();
+
+  pipeDesc.GS.BytecodeLength = 0;
+  pipeDesc.GS.pShaderBytecode = NULL;
+  pipeDesc.DS.BytecodeLength = 0;
+  pipeDesc.DS.pShaderBytecode = NULL;
+  pipeDesc.HS.BytecodeLength = 0;
+  pipeDesc.HS.pShaderBytecode = NULL;
+  pipeDesc.CS.BytecodeLength = 0;
+  pipeDesc.CS.pShaderBytecode = NULL;
+
+  ZeroMemory(&pipeDesc.RasterizerState, sizeof(D3D12_RASTERIZER_DESC));
+  pipeDesc.RasterizerState.FillMode = D3D12_FILL_MODE_SOLID;
+  pipeDesc.RasterizerState.CullMode = D3D12_CULL_MODE_NONE;
+  pipeDesc.RasterizerState.FrontCounterClockwise = TRUE;
+
+  pipeDesc.SampleDesc.Count = 1;
+  pipeDesc.SampleDesc.Quality = 0;
+
+  pipeDesc.InputLayout.NumElements = 0;
+  pipeDesc.InputLayout.pInputElementDescs = NULL;
+  pipeDesc.IBStripCutValue = D3D12_INDEX_BUFFER_STRIP_CUT_VALUE_DISABLED;
+  pipeDesc.PrimitiveTopologyType = D3D12_PRIMITIVE_TOPOLOGY_TYPE_TRIANGLE;
+  pipeDesc.RTVFormats.NumRenderTargets = 1;
+  pipeDesc.RTVFormats.RTFormats[0] = retFmt;
+  pipeDesc.DSVFormat = DXGI_FORMAT_UNKNOWN;
+
+  ID3D12PipelineState *samplePso = NULL;
+  HRESULT hr = m_pDevice->CreatePipeState(pipeDesc, &samplePso);
+  if(FAILED(hr))
+  {
+    RDCERR("Failed to create PSO for shader debugging HRESULT: %s", ToStr(hr).c_str());
+    SAFE_RELEASE(vsBlob);
+    SAFE_RELEASE(psBlob);
+    ShaderVariable lookupResult("tex", 0.0f, 0.0f, 0.0f, 0.0f);
+    output = lookupResult;
+    return true;
+  }
+
+  ID3D12GraphicsCommandListX *cmdList = m_pDevice->GetDebugManager()->ResetDebugList();
```

Comment:
You can release the blobs here right after creating the pipeline state, they're consumed inside that call and aren't used afterwards. Might make the error handling a little easier in the rest of the function not having to hold onto them.